### ISSUE TITLE
Option Proposal: for `no-unused-vars`, add `exceptions` and `exceptionsIgnorePattern` options

### ISSUE THREAD

**Issue by SpadeAceman** [eslint/eslint#3837]
Following from the discussion in #3543, I propose the addition of options to the `no-unused-vars` rule, allowing us to specify how we want to handle exception variables defined in the `catch()` clause of a try/catch statement.

Currently, the documentation doesn't address this, and the rule allows caught exception variables to remain unused, even when the `"all"` settings are used for the current options (tested via ESLint 1.4.3, though this behavior was also present in previous versions):
```
/*eslint no-unused-vars: [2, {"vars": "all", "args": "all"}]*/
function example(){
  try {
    doSomething();
  } catch (err) { // unused exception variable isn't flagged as an error
    return false;
  }
  return true;
}
```
Therefore, I propose the addition of `exceptions` and `exceptionsIgnorePattern` options, mirroring the functionality of the `vars`/`varsIgnorePattern` and `args`/`argsIgnorePattern` options. Valid settings for `exceptions` would be:

* `none` - the default, matches the current behavior of not reporting any unused exception variables
* `all` - reports all unused exception variables, aside from those matching `exceptionsIgnorePattern`
```
/*eslint no-unused-vars: [2, {"exceptions": "all", "exceptionsIgnorePattern": "^ignore"}]*/
function hypotheticalExample(){
  try {
    doSomething();
  } catch (err) { // unused exception variable IS flagged as an error
    return false;
  }
  try {
    doSomethingElse();
  } catch (ignoreErr) { // unused exception variable matches ignore pattern, is NOT flagged as an error
    return false;
  }
  return true;
}
```

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Looks good, let's do it.

**Comment by gyandeeps** [comment]
@nzakas Should we just make the default behavior to flag args inside catch statements and then also add `exceptionsIgnorePattern` option to the rule. We can release this with 2.0 with under the breaking section.

**Comment by nzakas** [comment]
I don't see a reason to introduce a breaking change for this. It flags things inside `catch` it just didn't flag the exception because it's required.

**Comment by mischah** [comment]
I know you hate those `+1` comments. 

… but I love to see this :kissing_heart:

**Comment by platinumazure** [comment]
@mischah Issue is accepted, so if you want to write a pull request, have at it! It's like somebody said (I forget who, maybe Gandhi?), "Be the change you wish to see in the world". So he probably would have also said "If you want something, write it yourself" :grin:

**Comment by pvamshi** [comment]
I am working on this , will create a pull request for review shortly

**Comment by ryanbriscall** [comment]
Thanks for this feature, but I needed this for function names.  It works for variables, given the provided "catch (ignoreErr)" example.  But doesn't work for functions, say for example: function ignoreTest() { }.  I was expecting this "no-unused-vars" exceptions solution to work for functions too, since eslint is showing my unused function error as a "no-unused-vars" error.
